compiler/rustc_hir_analysis/src/check/compare_method.rs

```diff
@@ -342,6 +346,31 @@ fn compare_predicate_entailment<'tcx>(
     Ok(())
 }
 
+fn compare_asyncness<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    impl_m: &AssocItem,
+    impl_m_span: Span,
+    trait_m: &AssocItem,
+    trait_item_span: Option<Span>,
+) -> Result<(), ErrorGuaranteed> {
+    if tcx.asyncness(trait_m.def_id) == hir::IsAsync::Async {
+        if tcx.asyncness(impl_m.def_id) != hir::IsAsync::Async {
```

Comment:
Why not make it both ways?

Reply:
There's not really a problem with using an async impl if the trait is `-> impl Future + '_`. I think it's not possible to create a refinement hazard with an async fn, but even if it is, that's already a problem with RPITIT that'll have to be solved before it can be stabilized, and I expect the solution to that to automatically cover whatever hypothetical refinement problem this might expose.

And we don't have to get it right in order to stabilize `async_fn_in_trait` since the RPITIT stuff is behind a different feature gate.